### PR TITLE
[HZ-1208] merge persistence & hot-restart when getting MapConfig

### PR DESCRIPTION
**Pull request by vbekiaris**
Merge persistence & hot-restart when getting MapConfig; while this is missing, map configs based on default or wildcard configs will not have their `DataPersistenceConfig` & `HotRestartConfig` merged, resulting in unexpected EE-side persistence behaviour (see [HZ-1208](https://hazelcast.atlassian.net/browse/HZ-1208)).

Additionally, when looking up config via
`ConfigUtils#lookupByPattern`, configs must be cloned and named
after the looked-for name, similarly to `ConfigUtils#getConfig`.
Otherwise, returned config has the name of the template config (eg "default" or "persistent.*") and this results in strange failures with persistence.

With these fixes, it is revealed that assertions in `ConfigSearchTest` and `DynamicConfigSearchOrderTest` are faulty, since they are based on config name. The tests are now fixed to compare on other properties of the config (eg backupCount).

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/5336 (only adds test on EE side)

Breaking changes (list specific methods/types/messages):
`Hazelcast.getConfig().findMapConfig("does-not-exist").getName()` (and similar `findXXX` methods) would previously (erroneously) return `default` instead of the name of the data structure that was requested.

Checklist:
- [ ] Label `Add to Release Notes` or `Not Release Notes content` set
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

**Comment by vbekiaris** [comment]
The fixes are in [DynamicConfigurationAwareConfig](https://github.com/hazelcast/hazelcast/pull/22355/files#diff-a46622d71610a9d796b71e206c1aff56087939ee8639693757a5abfddfada71cR268) & [ConfigUtils](https://github.com/hazelcast/hazelcast/pull/22355/files#diff-50dbca5dc4eb06e68d3442fcc744b0f3dc0c99e38dfd486759cfa0b42e43d1b0)

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigUtils.java [hunk 1]
```diff
@@ -44,7 +44,18 @@ public final class ConfigUtils {
     private ConfigUtils() {
     }
 
-    public static <T> T lookupByPattern(ConfigPatternMatcher configPatternMatcher, Map<String, T> configPatterns,
+    /**
+     * Lookup {@code itemName} in {@code configPatterns}, according to matching rules applied by
+     * {@code configPatternMatcher}.
+     * <p/>
+     * Probably this is NOT the method you are looking for: it returns the object located in {@code configPatterns}
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/internal/config/ConfigUtils.java:51]: obi-wan-kenobi-these-are-not-the-droids

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigUtils.java [hunk 2]
```diff
@@ -60,6 +71,40 @@ public static <T> T lookupByPattern(ConfigPatternMatcher configPatternMatcher, M
         return null;
     }
 
+    public static <T> T lookupByPattern(ConfigPatternMatcher configPatternMatcher,
+                                        @Nonnull Map<String, T> configPatterns,
+                                        String itemName, Class clazz) {
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/internal/config/ConfigUtils.java:76]: nit: clazz can be parametrized with `T` or `? extends T`

**Comment by frant-hartm** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
This PR is now out of date. A fix for HZ-1208 was included in 5.2 release. The clone-and-rename config issue should be addressed in a separate PR.

**Pull request closed without merging**